libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -112,8 +111,8 @@ bool VarNameCleaner::isUsedName(YulString const& _name) const
 {
 	if (_name.empty() || m_dialect.builtin(_name) || m_usedNames.count(_name))
```

Comment:
We already have a check for the builtin here. Is there anything in `Parser::instructions()` that is not allowed as name?

Reply:
@chriseth @ekpyron true!

There is actually a difference: `instructions()` includes all except `JUMPDEST` and `PUSHnn`, while `builtin` further removes `JUMP`, `JUMPI`, `DUPnn`, `SWAPnn`, and any opcode not present in the current EVM version.

Reply:
I am not sure what was the original motivation of this in the steps. One could specify `jump` as a function because the EVMDialect does not include it, but the VarNameCleaner/NameDispenser would consider it a special name and not touch it.

Reply:
Should we actually consider restricted names in dialects (e.g. identifiers which are not valid but also cannot be defined, such as `jump`, `jumpi` in EVM inline assembly)?

Reply:
Haha, actually the var name cleaner breaks for e.g. ``let if_1 := 0``, doesn't it :-)?

Reply:
Yep, I would actually prefer excluding tokens over excluding things like `jumpi`

Reply:
My question regarding `jump`/`jumpi` was more broad: I don't care if they are renamed or not in the optimiser, but should not be offered to the user.

Reply:
If they're not offered to the user, i.e. if they are error, then we have to exclude the renaming here, s.t. ``jump_1`` won't be renamed to ``jump``, as that'd be an error...
I would tend to agree with @chriseth, though, and don't consider it vital to disallow things like ``jump`` as yul identifiers...

Reply:
> I would tend to agree with @chriseth, though, and don't consider it vital to disallow things like jump as yul identifiers...

```
contract C {
    function f() {
       assembly { 
           function jump(a) { ... }
           ... one page of code here so the definition is not easy to see ...
           jump(next)
       }
    }
}
```

Is this something we should allow and we are okay with underhanded options or just not allow it (as it is a low hanging fruit)? I would only disallow EVM instructions which are not supported in Yul to avoid misleading code.

Reply:
So a dialect would have a set of reserved words?

Reply:
Yes.

Reply:
Introducing reserved keywords is tracked in #9330 and disallowing keywords is in #9329, because of that I think this PR should/could be merged as is.